zipkin/src/main/java/zipkin2/internal/ReadBuffer.java

```diff
@@ -0,0 +1,384 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zipkin2.internal;
+
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static zipkin2.internal.HexCodec.HEX_DIGITS;
+import static zipkin2.internal.JsonCodec.UTF_8;
+
+/** Read operations do bounds checks, as typically more errors occur reading than writing. */
+public abstract class ReadBuffer extends InputStream {
+
+  /** Do not use the buffer passed here after, as it may be manipulated directly. */
+  public static ReadBuffer wrapUnsafe(ByteBuffer buffer) {
+    if (buffer.hasArray()) {
+      int offset = buffer.arrayOffset() + buffer.position();
+      int limit = offset + buffer.remaining();
+      return wrap(buffer.array(), offset, limit);
+    }
+    return new ReadBuffer.Buff(buffer);
+  }
+
+  public static ReadBuffer wrap(byte[] bytes) {
+    return wrap(bytes, 0, bytes.length);
+  }
+
+  public static ReadBuffer wrap(byte[] bytes, int pos, int limit) {
+    return new ReadBuffer.Array(bytes, pos, limit);
+  }
+
+  static final class Buff extends ReadBuffer {
+    final ByteBuffer buf; // visible for testing
+
+    Buff(ByteBuffer buf) {
+      this.buf = buf;
+    }
+
+    @Override final byte readByteUnsafe() {
+      return buf.get();
+    }
+
+    @Override final byte[] readBytes(int length) {
+      require(length);
+      byte[] copy = new byte[length];
+      buf.get(copy);
+      return copy;
+    }
+
+    @Override boolean tryReadAscii(char[] destination, int length) {
+      buf.mark();
+      for (int i = 0; i < length; i++) {
+        byte b = buf.get();
+        if ((b & 0x80) != 0) {
+          buf.reset();
+          return false;  // Not 7-bit ASCII character
+        }
+        destination[i] = (char) b;
+      }
+      return true;
+    }
+
+    @Override final String doReadUtf8(int length) {
```

Comment:
For reference, when working on https://github.com/protocolbuffers/protobuf/pull/3824 I couldn't get a `CharsetDecoder` to be faster than copying into an array like here. I can't remember if I tried using a threadlocal though to reuse the decoder.

Anyways, at some point we could think about whether it makes sense to copy over some Unsafe code to read the bytes directly.

Reply:
makes sense